Decouple tuple

deinst opened the pull request:
Change sage/combinat/tuple.py to use itertools instead of GAP.

### :books: Description

This fixes #35784.
Iterating Tuples delegates to itertools.product
Iterating UnorderedTuples delegates to itertools.combinations_with_replacement

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. It should be `[x]` not `[x ]`. -->

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies


<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

deinst committed:
Mind the GAP

The generators for the classes Tuples and UnorderedTuples used
the GAP for generating Tuples.  This could be done much more simply
using itertools.

Also fixed a bug where if the underlying list had multiple identical
elements then redundant tuples were generated.

deinst committed:
Finish decoupling tuple.py from GAP.

This commit is just a cosmetic change from the last (deleting a
commented out line).  This should finish bug #35784.

deinst committed:
Added testcases to Tuples for undelying lists with duplicates

Also removed a blank line that irritated the linter.

dcoudert commented on src/sage/combinat/tuple.py:
```diff
@@ -75,7 +75,7 @@ def __init__(self, S, k):
         self.S = S
         self.k = k
-        self._index_list = [S.index(s) for s in S]
+        self._index_list = list(dict.fromkeys(S.index(s) for s in S))
```
Why not simply using `self._index_list = list(range(len(S)))` ?

dcoudert replied:
may be `self._index_list = range(len(S))` is enough

deinst replied:
We need to remove duplicates from the list (that is the  bug noted in the conversation in the bug report).  The reason that I use the `dict.fromkeys` dance instead of `set` is so that the order of elements is not changed, leaving the order of the generated tuples the same as before.  (The same is true of the seemingly gratuitous `reverse` in the `__iter__`)

The reason for the `_index_list` is to remove duplicates from a list that has non-hashable elements.

range(len(S)) absolutely won't work.

dcoudert replied:
I see. Thanks for the precise answer.

tscrim replied:
I think it is very bad to rely on the `dict` order behavior (which could change without notice) and it is not clear that this is what the code is trying to do. IMO it would be much better to then run `sorted(set(self._index_list))` to get the unique elements.

deinst replied:
I'm more than willing to go to using `set`, it certainly is clearer.   The fact that the order is preserved by dict has been guaranteed since python 3.7.

dcoudert replied:
The drawback of going to `set` is to no longer accept non-hashable elements...

deinst replied:
We still do the index list dance so we are just hashing integers regardless.  All using `dict` bought us was a guarantee of preserving the order at the cost of some clarity in the code.

tscrim replied:
They are just `int`s. Please read what I actually wrote more carefully.

Edit: Perhaps the fact that I was basing everything off the code before the change was unclear; so `self._index_list = [S.index(s) for s in S]`.

tscrim replied:
I know the order is currently guaranteed, but that is basically an implementation detail that could change (as it is not the fundamental programming model for a `dict`). It also makes for very brittle code a change the insertion order means everything gets broken.

mkoeppe replied:
The order is guaranteed precisely so that one can rely on it in code.

tscrim replied:
Even if there is now a strong guarantee that it won't change (which I would not say there is considering Python has shown it is okay breaking backwards incompatibility), it still is brittle and obfuscated code. Having an explicit `sorted` makes the intent of the code clear.

mkoeppe replied:
> considering Python has shown it is okay breaking backwards incompatibility

If you're referring to the Python 2 -> Python 3 transition, well that won't happen again.

dcoudert commented on src/sage/combinat/tuple.py:
```diff
@@ -178,7 +167,7 @@ def __init__(self, S, k):
         self.S = S
         self.k = k
-        self._index_list = [S.index(s) for s in S]
+        self._index_list = list(dict.fromkeys(S.index(s) for s in S))
```
same here.

dcoudert reviewed (approved):
LGTM.

deinst committed:
Have tuple.py return tuples instead of lists

This has Tuples and UnorderedTuples return tuples.
Need to fix up schemes/projective/projective_space.py as it assumes the
tuples are lists.

dcoudert commented on src/sage/schemes/projective/projective_space.py:
```diff
@@ -2374,7 +2374,7 @@ def rational_points(self, bound=0):
                 P[i] = ai
                 for tup in S[i - 1]:
-                    if gcd([ai] + tup) == 1:
+                    if gcd([ai] + list(tup)) == 1:
```
Here, `gcd((ai,) + tup)` is more appropriate.
```py
sage: def A(a, tup):
....:     return [a] + list(tup)
....: 
sage: def B(a, tup):
....:     return (a,) + tup
....: 
sage: tup = tuple(range(10))
sage: %timeit A(100, tup)
134 ns ± 0.0589 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
sage: %timeit B(100, tup)
84.8 ns ± 0.0729 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
sage: tup = tuple(range(4))
sage: %timeit A(100, tup)
123 ns ± 1.49 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
sage: %timeit B(100, tup)
79.1 ns ± 0.0314 ns per loop (mean ± std. dev. of 7 runs, 10,000,000 loops each)
sage: %timeit gcd(A(100, tup))
7.03 µs ± 4.32 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
sage: %timeit gcd(B(100, tup))
6.95 µs ± 11.7 ns per loop (mean ± std. dev. of 7 runs, 100,000 loops each)
```

tscrim replied:
That's not a fair comparison because of the extra (unnecessary) `list` cast, which creates a copy. (Recall the old code was returning a list.)

dcoudert replied:
> That's not a fair comparison because of the extra (unnecessary) `list` cast, which creates a copy. (Recall the old code was returning a list.)
I agree. Since the code now returns a tuple, it's better to use `(ai,) + tup` than casting to list.

deinst committed:
Switch tuple to list in projective_space.py

Part of the ongoing project to free Tuple of GAP.

dcoudert reviewed (approved):
LGTM.

deinst committed:
Merge branch 'develop' into decouple_tuple

deinst committed:
Change dict to set

The use of the python dict guarantee of insertion order of keys was
confusing.  There was no compelling reason (other than an aversion to
change) for keeping insertion order.

tscrim reviewed (approved):
LGTM. Thanks.

Commit:
Merge tag '10.1.beta5' into decouple_tuple

SageMath version 10.1.beta5, Release Date: 2023-07-01

dcoudert reviewed (approved):
LGTM.
